library/core/src/task/wake.rs

```diff
@@ -382,6 +388,13 @@ impl Clone for Waker {
             waker: unsafe { (self.waker.vtable.clone)(self.waker.data) },
         }
     }
+
+    #[inline]
+    fn clone_from(&mut self, source: &Self) {
+        if !self.will_wake(source) {
+            *self = source.clone();
+        }
+    }
```

Comment:
Normally, Waker runs clone side effects. This, in some cases, avoids doing so. This seems like observable behavior. Are we okay with changing it? Are there implementations "in the wild" that will be affected negatively by this? Normally, `clone_from` is opt-in, so I think it's totally reasonable to have it differ slightly from the naive case of  `clone`... this _is_ its stated purpose, after all.

Reply:
This changes how the code is executed, but not its semantics. Like `String::clone_from` can avoid allocating, but the semantic result compared to `*self = other.clone()` is the same.

Moreover, I can't imagine any sane code where this breaks some code.

Reply:
It can potentially have an arbitrary side-effect, like printing an integer, but of course I think ruling that as a fundamentally unserious use-case (or at least, easily circumvented by not calling `clone_from`) seems reasonable.